File: 백준/Gold/3079. 입국심사/입국심사.cc
#include <algorithm>
#include <cstdio>

using ll = long long;

int n, v[100001];
ll m, s;
bool clc(const ll& tim) {
  s = 0;
  for (int i = 0; i < n && s <= m; i++) s += (tim / v[i]);
  return s >= m;
}

ll sol() {
  ll beg = 1, end = m * v[n - 1], mid, res = m * v[n - 1];
  while (beg <= end) {
    mid = (beg + end) / 2;
    if (clc(mid)) {
      res = std::min(mid, res);
      end = mid - 1;
    } else {
      beg = mid + 1;
    }
  }
  return res;
}

int main() {
  scanf("%d %lld", &n, &m);
  for (int i = 0; i < n; i++) scanf("%d", &v[i]);
  std::sort(v, v + n);
  printf("%lld", sol());
}

File: 프로그래머스/lv3/43238. 입국심사/입국심사.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public boolean sol(final long tim, final int n, final int[] v) {
        return IntStream.range(0, v.length)
            .mapToLong(i -> tim / v[i])
            .sum() >= n;
    }
    
    public long solution(int n, int[] times) {
        Arrays.sort(times);
        long answer = (long) times[times.length - 1] * n;
        long s = 1, e = (long) times[times.length - 1] * n;
        while(s <= e) {
            long mid = (s + e) / 2;
            if (sol(mid, n, times)) {
                answer = Math.min(answer, mid);
                e = mid - 1;
            } else {
                s = mid + 1;
            }
        }
        return answer;
    }
}